### PR TITLE
Add &self arg to intern_param

### PR DESCRIPTION
**Pull request by yaahc**
related to #340

### CODE REVIEW

File: chalk-ir/src/fold/boring_impls.rs [hunk 1]
```diff
@@ -71,25 +71,26 @@ impl<I: Interner, TI: TargetInterner<I>> Fold<I, TI> for Parameter<I> {
...
         let data = self.data().fold_with(folder, binders)?;
-        Ok(Parameter::new(data))
+        Ok(Parameter::new(folder.target_interner(), data))
     }
 }
 
 impl<I: Interner, TI: TargetInterner<I>> Fold<I, TI> for Substitution<I> {
     type Result = Substitution<TI>;
     fn fold_with(&self, folder: &mut dyn Folder<I, TI>, binders: usize) -> Fallible<Self::Result> {
+        let folded: Vec<_> = self.iter().map(|p| p.fold_with(folder, binders)).collect();
```
Comment (yaahc) [inline chalk-ir/src/fold/boring_impls.rs]: Had to collect here and blow to deal with lifetimes. `p.fold_with` takes a &mut ref to the folder and `target_interner()` takes a & ref to the folder.

File: chalk-derive/src/lib.rs [hunk 2]
```diff
@@ -30,11 +30,14 @@ pub fn derive_fold(item: TokenStream) -> TokenStream {
...
 
-                fn fold_with(
+                fn fold_with<'i>(
                     &self,
-                    folder: &mut dyn Folder < #arg, #arg >,
+                    folder: &mut dyn Folder < 'i, #arg, #arg >,
                     binders: usize,
-                ) -> ::chalk_engine::fallible::Fallible<Self::Result> {
+                ) -> ::chalk_engine::fallible::Fallible<Self::Result>
+                where
```
Comment (nikomatsakis) [inline chalk-derive/src/lib.rs]: Nit: this diff is probably not needed, but I guess it doesn't hurt -- in particular `#arg` will normally be a type name like `ChalkIr`, so this is "clearly true"

File: chalk-ir/src/lib.rs [hunk 3]
```diff
@@ -954,14 +958,18 @@ pub enum ProgramClause<I: Interner> {
     ForAll(Binders<ProgramClauseImplication<I>>),
 }
 
+impl<I: Interner> HasInterner for ProgramClause<I> {
```
Comment (nikomatsakis) [inline chalk-ir/src/lib.rs]: Pretty sure we can `#[derive]` this

File: chalk-ir/src/lib.rs [hunk 4]
```diff
@@ -983,6 +991,13 @@ pub struct Canonical<T> {
     pub binders: Vec<ParameterKind<UniverseIndex>>,
 }
 
+impl<T> HasInterner for Canonical<T>
```
Comment (nikomatsakis) [inline chalk-ir/src/lib.rs]: Similarly, I suspect we can `#[derive]` this

**Pull request merged**